drivers/SmartThings/zigbee-thermostat/src/init.lua

```diff
@@ -164,6 +169,61 @@ local thermostat_mode_setter = function(mode_name)
   end
 end
 
+
+local emit_cool_setpoint_bounds_event = function (device)
+  device:emit_event(capabilities.thermostatCoolingSetpoint.coolingSetpointRange(
+      {
+        unit = 'C',
+        value = {minimum = device:get_field(SETPOINT_MIN_COOL),
+                 maximum = device:get_field(SETPOINT_MAX_COOL)}
+      }
+    )
+  )
+  device:set_field(SETPOINT_MIN_COOL, nil)
+  device:set_field(SETPOINT_MAX_COOL, nil)
+end
+
+local emit_heat_setpoint_bounds_event = function (device)
+  device:emit_event(capabilities.thermostatHeatingSetpoint.heatingSetpointRange(
+      {
+        unit = 'C',
+        value = {minimum = device:get_field(SETPOINT_MIN_HEAT),
+                 maximum = device:get_field(SETPOINT_MAX_HEAT)}
+      }
+    )
+  )
+  device:set_field(SETPOINT_MIN_HEAT, nil)
+  device:set_field(SETPOINT_MAX_HEAT, nil)
+end
+
+local thermostat_setpoint_min_cool_handler = function(driver, device, setpoint)
+  device:set_field(SETPOINT_MIN_COOL, setpoint.value)
+  if device:get_field(SETPOINT_MIN_COOL) and device:get_field(SETPOINT_MAX_COOL) then
+      emit_cool_setpoint_bounds_event(device)
+  end
+end
+
+local thermostat_setpoint_max_cool_handler = function(driver, device, setpoint)
+  device:set_field(SETPOINT_MAX_COOL, setpoint.value)
+  if device:get_field(SETPOINT_MIN_COOL) and device:get_field(SETPOINT_MAX_COOL) then
+      emit_cool_setpoint_bounds_event(device)
+  end
+end
+
+local thermostat_setpoint_min_heat_handler = function(driver, device, setpoint)
+  device:set_field(SETPOINT_MIN_HEAT, setpoint.value)
+  if device:get_field(SETPOINT_MIN_HEAT) and device:get_field(SETPOINT_MAX_HEAT) then
+      emit_heat_setpoint_bounds_event(device)
+  end
+end
+
+local thermostat_setpoint_max_heat_handler = function(driver, device, setpoint)
+  device:set_field(SETPOINT_MAX_HEAT, setpoint.value)
+  if device:get_field(SETPOINT_MIN_HEAT) and device:get_field(SETPOINT_MAX_HEAT) then
+      emit_heat_setpoint_bounds_event(device)
+  end
+end
```

Comment:
You could save yourself some repetition here if you wanted by using a factory for creating functions here. Something like:

```
local setpoint_limit_handler_factory = function(handled_attribute, paired_attribute)
  return function(driver, device, setpoint)
    device:set_field(handled_attribute, setpoint.value)
    if device:get_field(paired attribute) and device:get_field(handled_attribute) then
      emit_heat_setpoint_bounds_event(device)
    end
  end
end
```

Then in the handlers below you could do

```
[Thermostat.attributes.MinHeatSetpointLimit.ID] = setpoint_limit_handler_factory(SETPOINT_MIN_HEAT, SETPOINT_MAX HEAT),
```

Reply:
Done. I created a handler factory.